core/runtime/impl/grandpa_impl.cpp

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "runtime/impl/grandpa_impl.hpp"
+
+#include "runtime/impl/wasm_memory_stream.hpp"
+#include "scale/collection.hpp"
+#include "scale/optional.hpp"
+#include "scale/scale_error.hpp"
+
+namespace kagome::runtime {
+  using primitives::Digest;
+  using primitives::ForcedChangeType;
+  using primitives::ScheduledChange;
+  using primitives::SessionKey;
+  using primitives::WeightedAuthority;
+
+  GrandpaImpl::GrandpaImpl(common::Buffer state_code,
+                           std::shared_ptr<extensions::Extension> extension,
+                           std::shared_ptr<primitives::ScaleCodec> codec)
+      : memory_(extension->memory()),
+        codec_(std::move(codec)),
+        executor_(std::move(extension)),
+        state_code_(std::move(state_code)) {}
+
+  outcome::result<std::optional<ScheduledChange>> GrandpaImpl::pendingChange(
+      const Grandpa::Digest &digest) {
+    OUTCOME_TRY(encoded_digest, codec_->encodeDigest(digest));
+
+    runtime::SizeType ext_size = encoded_digest.size();
+    runtime::WasmPointer ptr = memory_->allocate(ext_size);
```

Comment:
As Pavel noticed, it would be good to leave a TODO here concerning memory overflow check.

Reply:
I believe there is already a task for it, so you can reuse it.

Reply:
added TODO